osquery/core/query.cpp

```diff
@@ -187,20 +204,14 @@ Status Query::addNewResults(QueryDataTyped current_qd,
       return status;
     }
 
-    status = setDatabaseValue(kQueries, name_, json);
-    if (!status.ok()) {
-      return status;
-    }
-
-    status = setDatabaseValue(
-        kQueries, name_ + "epoch", std::to_string(current_epoch));
+    status = saveQueryResults(json, current_epoch);
     if (!status.ok()) {
       return status;
     }
   }
 
-  if (update_db || fresh_results || new_query) {
-    auto status = incrementCounter(fresh_results || new_query, counter);
+  if (update_db || new_epoch || new_query) {
+    auto status = incrementCounter(new_epoch, new_query, counter);
```

Comment:
Why not `    auto status = incrementCounter(true, new_epoch || new_query, counter);`

Will this do the right thing with the same epoch and a new query?

Reply:
The first argument should be `false` if it's the same epoch because that means the results are a diff, rather than including all records (even if the query changes). In the same epoch but new query case, counter resets to `1` instead of `0`.

`auto status = incrementCounter(new_epoch, new_epoch || new_query, counter)` would work here though. It won't have different behavior since new_epoch always forces counter of `0` anyway, happy to change if you think it's clearer.

Reply:
I am very confused here. 

I _thought_ `new_epoch` referred to a new epoch, eg, the epoch has been changed. And that `new_query` referred to the first execution of a query. (I get this from reading the comments in the code).

As I understand the goal, either a new_query OR a new_epoch should reset the counter. And then what we reset the counter to depends on whether it's evented or not. Put differently, I think if there's a new_epoch, _every_ query should be a new_query. And that we should be calling this as `incrementCounter(<event-or-regular>, new_epoch ||  new_query, counter);`

Reply:
Ah yes I see the confusion, I had been using "new query" to mean the query SQL changed, not necessarily when the query is brand new overall. And "new epoch" to mean a new epoch for this query, including its first epoch overall.

I think it could be clearer if I rename these variables instead to `new_query_epoch` and `query_sql_changed` -- I'll try that now but happy to revert or tweak if you prefer other names.

Here's another way to describe the behaviors I was going for in this PR:

_non-evented table_
1. query is brand new -> reset counter to 0, include all records
2. epoch changed -> reset counter to 0, include all records
3. epoch same, query sql changed -> **reset counter to 1** (formerly 0), do a normal diff
4. epoch same, query sql same -> increment counter by 1, do a normal diff

_evented table_
1. query is brand new -> **reset counter to 1** (formerly 0), stream events
2. epoch changed -> **reset counter to 1** (formerly 0), stream events
3. epoch same, query sql changed -> **reset counter to 1** (formerly 0), stream events
4. epoch same, query sql same -> increment counter by 1, stream events

The bolded parts are what this PR changes vs what osquery does today. To me the new behavior better matches the intent of the documentation to be able "to skip the initial `added` records" by ignoring when counter = 0 described [here](https://osquery.readthedocs.io/en/stable/deployment/logging/#schedule-counter).

Reply:
(updated the PR, hopefully the naming makes more sense!)

Reply:
Oh, your table form is helpful. I am confused, because I disagree with one of the lines:
> non-evented table
>
> 1. query is brand new -> reset counter to 0, include all records
> 2. epoch changed -> reset counter to 0, include all records
> 3. epoch same, query sql changed -> reset counter to 1 (formerly 0), do a normal diff
> 4. epoch same, query sql same -> increment counter by 1, do a normal diff

I think (1), (2), and (3) should have the same behavior. They should reset to 0 and includes all records.

I think a SQL change is functionally a new query. I'm curious why not?

Reply:
Wait, maybe we're both wrong?

If the query name is unchanged, we should _not_ be resetting the the counter. `[query name, epoch, counter]` _should_ be unique. Changing the query sql, but not the name, should not change the counter because it would invalidate that. It's also probably a bad idea.

Reply:
Yeah I think there are good arguments either way for whether to return all records for a SQL change, so I was hesitant to change the current behavior of it doing a diff - it seemed like a larger/potentially more disruptive change than changing the counter behavior.

> If the query name is unchanged, we should not be resetting the the counter.

I like this idea! While I was most bothered by the counter being 0 in this case, I'm not actually sure why resetting at all there is helpful -- I agree violating the uniqueness of that triple is annoying. 

I'm happy to make this update. If anyone really relied on knowing when results are due to a query SQL update for some reason, in theory they could do that in a more clear/direct way by looking for the status line or joining on the `osquery_schedule` table in the query itself.

Reply:
> Yeah I think there are good arguments either way for whether to return all records for a SQL change, so I was hesitant to change the current behavior of it doing a diff - it seemed like a larger/potentially more disruptive change than changing the counter behavior.

Yes, I think that's the correct judgement -- This PR should preserve the existing behavior. Not that we can't fix it, but we should be deliberate about doing so.

In that vein, this PR feels reasonable. It's not a deep change. Can you open a blueprint issue with a proposal for better counter behavior on changing SQL? I imagine there might be discussion to be had (I can think of 2 solutions)

Reply:
Sounds good! I opened one here https://github.com/osquery/osquery/issues/7945